hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/impl/AbstractGrantRevokeKycPrecompile.java

```diff
@@ -0,0 +1,118 @@
+import com.hedera.services.store.contracts.precompile.codec.DecodingFacade;
+import com.hedera.services.store.contracts.precompile.codec.GrantRevokeKycWrapper;
+import com.hedera.services.store.contracts.precompile.utils.KeyActivationUtils;
+import com.hedera.services.store.contracts.precompile.utils.PrecompilePricingUtils;
+import com.hedera.services.store.models.Id;
+import com.hedera.services.txns.token.GrantKycLogic;
+import com.hedera.services.txns.token.RevokeKycLogic;
+import com.hederahashgraph.api.proto.java.ResponseCodeEnum;
+import java.util.Objects;
+import org.hyperledger.besu.evm.frame.MessageFrame;
+
+public abstract class AbstractGrantRevokeKycPrecompile extends AbstractWritePrecompile {
+    private final boolean hasGrantKycLogic;
+    protected final ContractAliases aliases;
+    protected final EvmSigsVerifier sigsVerifier;
+    protected GrantRevokeKycWrapper grantRevokeOp;
+
+    protected AbstractGrantRevokeKycPrecompile(
+            WorldLedgers ledgers,
+            DecodingFacade decoder,
+            ContractAliases aliases,
+            EvmSigsVerifier sigsVerifier,
+            SideEffectsTracker sideEffects,
+            SyntheticTxnFactory syntheticTxnFactory,
+            InfrastructureFactory infrastructureFactory,
+            PrecompilePricingUtils pricingUtils,
+            boolean hasGrantKycLogic) {
+        super(
+                ledgers,
+                decoder,
+                sideEffects,
+                syntheticTxnFactory,
+                infrastructureFactory,
+                pricingUtils);
+        this.aliases = aliases;
+        this.sigsVerifier = sigsVerifier;
+        this.hasGrantKycLogic = hasGrantKycLogic;
+    }
+
+    @Override
+    public void run(MessageFrame frame) {
+        Objects.requireNonNull(grantRevokeOp);
+
+        /* --- Check required signatures --- */
+        final var tokenId = Id.fromGrpcToken(grantRevokeOp.token());
+        final var accountId = Id.fromGrpcAccount(grantRevokeOp.account());
+        final var hasRequiredSigs =
+                KeyActivationUtils.validateKey(
+                        frame,
+                        tokenId.asEvmAddress(),
+                        sigsVerifier::hasActiveKycKey,
+                        ledgers,
+                        aliases);
+        validateTrue(hasRequiredSigs, INVALID_SIGNATURE);
+
+        /* --- Build the necessary infrastructure to execute the transaction --- */
+        final var accountStore = infrastructureFactory.newAccountStore(ledgers.accounts());
+        final var tokenStore =
+                infrastructureFactory.newTokenStore(
+                        accountStore,
+                        sideEffects,
+                        ledgers.tokens(),
+                        ledgers.nfts(),
+                        ledgers.tokenRels());
+
+        /* --- Execute the transaction and capture its results --- */
+        if (hasGrantKycLogic) {
```

Comment:
This logic I've seen in several precompile pairs and it still has not felt right to me.  We have specialized logic based on the subclass, but instead of putting the logic in the subclass we set a boolean on the parent and switch based on the boolean.  This should be a call to a virtual method where the subclass does the logic.  If we are specializing on behavior the behavior belongs in the object that is doing the specializing, not the parent.  Otherwise we should just have a factory method that spits out this method and sets the params based on the call.  This weird hybrid of both reduces readability.

Reply:
DONE